src/cluster.c

```diff
@@ -1042,8 +1041,19 @@ static dict *parse_cluster_nodes(valkeyClusterContext *cc, valkeyReply *reply) {
 error:
     sdsfreesplitres(part, count_part);
     sdsfreesplitres(slot_start_end, count_slot_start_end);
+    if (nodes_name != NULL) {
+        /* Only free parsed replicas since the `nodes` dict owns primary nodes. */
```

Comment:
This is a little odd and maybe fragile.

What about the corner case primaries without any slots? Does the `nodes` dict own them too?

Reply:
It is a bit odd.
The `nodes` dict owns all `valkeyClusterNodes` after the parsing is complete, these are nodes with-or-without slots. This dict holds key=address, value=primary node, and each primary owns a list of replicas.
Deleting the `nodes` dict will delete all its objects, primaries and their known replicas, when there is an error while parsing.

But then there is the temporary `nodes_name` dict which holds references to all currently parsed cluster nodes.
The found issue is that this dict is the actual owner of an replica node, that has a primary, which is not yet parsed.
So, these are the replica nodes that needs to be deleted while the parsing fails.

I have a coming PR where the cluster nodes parsing is refactored.